### ISSUE TITLE
AMP should raise MissingArgument (or other) if a callRemote is called with wrong arguments

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2808]
|<img alt="Lacrymology's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Lacrymology reported|
|-|-|
|Trac ID|trac#2808|
|Type|defect|
|Created|2007-08-31 14:27:50Z|
|Branch|https://github.com/twisted/twisted/tree/amp-wrong-args-2808|

if callRemote(SomeCommand,someArg="foo") (where someArg is optional) is called with a wrong argument, no exception is raised, and the optional argument reaches the server with a None value

Attachments:

* [test_wrongArguments.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bb9/bb979a33849ad72fa65bb94815dd92e08da9f2f1/876192c402269fa6f5e0fde857e81d8765cd3cf4.py) (953 bytes) - added by Lacrymology on 2007-08-31 15:08:57Z - This test fails because of this
* [patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bb9/bb979a33849ad72fa65bb94815dd92e08da9f2f1/d75b3b528276d7a9f30a04b8bccaf74e1d61f67a) (2432 bytes) - added by Lacrymology on 2007-08-31 15:59:39Z - patch
* [amp-parameter-enforcement.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bb9/bb979a33849ad72fa65bb94815dd92e08da9f2f1/c94a4455e0e0e6ad3871395d5eb4025ea8ff9467.diff) (1604 bytes) - added by exarkun on 2007-09-02 21:54:47Z - updated version of the test and implementation change
* [amp-parameter-enforcement-fixed.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bb9/bb979a33849ad72fa65bb94815dd92e08da9f2f1/d4539eb605d18dd0c803dac9fad734411fa30c86.diff) (1512 bytes) - added by Lacrymology on 2007-09-26 18:06:58Z - 
* [amp-wrong-parameter.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bb9/bb979a33849ad72fa65bb94815dd92e08da9f2f1/a52a487cd5d46c4a33a03793597b48499630d012.patch) (1522 bytes) - added by exarkun on 2007-10-31 20:33:35Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2808 2808
type__defect defect
reporter__Lacrymology Lacrymology
priority__highest highest
milestone__ 
branch__branches_amp_wrong_args_2808 branches/amp-wrong-args-2808
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__AMP AMP
time__1188570470000000 1188570470000000
changetime__1229395704000000 1229395704000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Lacrymology's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Lacrymology set status to `closed`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="Lacrymology's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Lacrymology set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="Lacrymology's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|Lacrymology removed owner<br>Lacrymology set status to `new`|
|-|-|

Attached a patch with my solution

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

I simplified the test a bit.  Attaching an updated patch.

Three existing amp unit tests are broken, though.  Could you take a look at those?

**Comment by twisted-trac** [comment]
|<img alt="Lacrymology's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|Lacrymology commented|
|-|-|

There, I've fixed the patch so it passes the argnames through _wireNameToPythonIdentifier before comparing

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to Lacrymology|
|-|-|

 * your patch contains tabs
 * the test doesn't test your change

**Comment by twisted-trac** [comment]
|<img alt="Lacrymology's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Lacrymology removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set owner to Lacrymology|
|-|-|

 * I think ```_wireNameToPythonIdentifier``` should be applied to ```intendedArg```, or there is a asymmetry I don't understand
 * ```InvalidSignature``` should be raised with an explicit message, like "%s argument not existing" or something

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="Lacrymology's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|Lacrymology set owner to @therve|
|-|-|

```intendedArg``` is in the ```PythonIdentifier``` "format", this is it's "Print" instead of 'print' and "foo_bar" instead of "foo-bar" because they've been created through keyword paramenter assignment. Since the convertion is not biyective (there's no way of knowing if "foo_bar" in the keywords dict is supposed to be an argument called "foo-bar" or one called "foo_bar". This is, the ```WireName``` realm is bigger than the ```PythonIdentifier``` one, so if you want to check that one corresponds to the other, you need to convert from bigger to smaller (yeay, Sets Algebra was good for something, even if I don't remember the terminology)

The patch has been fixed

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to Lacrymology|
|-|-|

If cls.arguments can contain python identifier, intendedArg can too. For example:

```
client.callRemote(Hello, **{"print": "print"})
```

**Comment by twisted-trac** [comment]
|<img alt="Lacrymology's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|Lacrymology set owner to @therve|
|-|-|

ok, then a not - or is in order.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to Lacrymology|
|-|-|

I reformatted a few lines of the change, and generated a diff which patch can actually apply.

There is a problem with the implementation - it checks if ```intendedArg``` is in ```cls.arguments``` or not, but ```cls.arguments``` is a list of two-tuples, so ```intendedArg``` will never be in it.  Also there's no test coverage for the case where this check is required.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [25640]) Branching to 'amp-wrong-args-2808'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [25641]) Apply amp-wrong-parameter.patch

refs [#2808](https://github.com/twisted/twisted/issues/2808)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [25642]) Make sure to test the interaction of python identifiers which aren't actually allowed to be passed to Command.makeResponse without being mangled

refs [#2808](https://github.com/twisted/twisted/issues/2808)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun removed owner|
|-|-|

Applied the patch to a branch, fixed the review comments I made above, please review.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_19"></a>|@washort set owner to @exarkun|
|-|-|

Looks good. Merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun set status to `closed`|
|-|-|

(In [25679]) Merge amp-wrong-args-2808

Author: lacrymology, exarkun
Reviewer: exarkun, washort
Fixes: [#2808](https://github.com/twisted/twisted/issues/2808)


Check for arguments in `Command.makeArguments` which are not defined in the
command's arguments list and reject them by raising `InvalidSignature`.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
